src/topology/noetherian_space.lean

```diff
@@ -0,0 +1,220 @@
+  rw map_sup,
+  convert hx',
+  exact sup_eq_left.mpr hx''
+end
+
+variable (α)
+
+example (α : Type*) : set α ≃o (set α)ᵒᵈ := by refine order_iso.compl (set α)
+
+lemma noetherian_space_tfae :
+  tfae [noetherian_space α,
+    well_founded (λ s t : closeds α, s < t),
+    ∀ s : set α, is_compact s,
+    ∀ s : opens α, is_compact (s : set α)] :=
+begin
+  tfae_have : 1 ↔ 2,
+  { refine (noetherian_space_iff _).trans (surjective.well_founded_iff opens.compl_bijective.2 _),
+    exact λ s t, (order_iso.compl (set α)).lt_iff_lt.symm },
+  tfae_have : 1 ↔ 4,
+  { exact noetherian_space_iff_opens α },
+  tfae_have : 1 → 3,
+  { intros H s, rw is_compact_iff_compact_space, resetI, apply_instance },
+  tfae_have : 3 → 4,
+  { exact λ H s, H s },
+  tfae_finish
+end
+
+variables {α β}
+
+lemma noetherian_space.is_compact [h : noetherian_space α] (s : set α) : is_compact s :=
+let H := (noetherian_space_tfae α).out 0 2 in H.mp h s
+
+lemma noetherian_space_of_surjective [noetherian_space α] (f : α → β)
+  (hf : continuous f) (hf' : function.surjective f) : noetherian_space β :=
+begin
+  rw noetherian_space_iff_opens,
+  intro s,
+  obtain ⟨t, e⟩ := set.image_surjective.mpr hf' s,
+  exact e ▸ (noetherian_space.is_compact t).image hf,
+end
+
+lemma noetherian_space_iff_of_homeomorph (f : α ≃ₜ β) :
+  noetherian_space α ↔ noetherian_space β :=
+⟨λ h, @@noetherian_space_of_surjective _ _ h f f.continuous f.surjective,
+  λ h, @@noetherian_space_of_surjective _ _ h f.symm f.symm.continuous f.symm.surjective⟩
+
+lemma noetherian_space.range [noetherian_space α] (f : α → β) (hf : continuous f) :
+  noetherian_space (set.range f) :=
+noetherian_space_of_surjective (set.cod_restrict f _ set.mem_range_self) (by continuity)
+  (λ ⟨a, b, h⟩, ⟨b, subtype.ext h⟩)
+
+lemma noetherian_space_set_iff (s : set α) :
+  noetherian_space s ↔ ∀ t ⊆ s, is_compact t :=
+begin
+  rw (noetherian_space_tfae s).out 0 2,
+  split,
+  { intros H t ht,
+    have := embedding_subtype_coe.is_compact_iff_is_compact_image.mp (H (coe ⁻¹' t)),
+    simpa [set.inter_eq_left_iff_subset.mpr ht] using this },
+  { intros H t,
+    refine embedding_subtype_coe.is_compact_iff_is_compact_image.mpr (H (coe '' t) _),
+    simp }
+end
+
+@[simp] lemma noetherian_univ_iff :
+  noetherian_space (set.univ : set α) ↔ noetherian_space α :=
+noetherian_space_iff_of_homeomorph (homeomorph.set.univ α)
+
+lemma noetherian_space.Union {ι : Type*} [fintype ι] (f : ι → set α)
+  [hf : ∀ i, noetherian_space (f i)] :
+  noetherian_space (⋃ i, f i) :=
+begin
+  simp_rw noetherian_space_set_iff at hf ⊢,
+  intros t ht,
+  rw [← set.inter_eq_left_iff_subset.mpr ht, set.inter_Union],
+  exact compact_Union (λ i, hf i _ (set.inter_subset_right _ _))
+end
+
+-- This is not an instance since it makes a loop with `t2_space_discrete`.
+lemma noetherian_space.discrete [noetherian_space α] [t2_space α] : discrete_topology α :=
+⟨eq_bot_iff.mpr (λ U _, is_closed_compl_iff.mp (noetherian_space.is_compact _).is_closed)⟩
```

Comment:
Isn't there a lemma that says `compact_space + t2_space -> discrete_topology`?

Reply:
I'm not sure. I couldn't find it.